Sources/Regex/Parse/Parse.swift

```diff
@@ -183,19 +215,56 @@ extension Parser {
       return .symmetricDifference
     }
   }
+*/
 
-  ///     CharacterClass -> '[' CharacterSetComponent+ ']'
+  /// Parse a custom character class
   ///
-  ///     CharacterSetComponent -> CharacterSetComponent SetOp CharacterSetComponent
-  ///     CharacterSetComponent -> CharacterClass
-  ///     CharacterSetComponent -> <token: Character>
-  ///     CharacterSetComponent -> <token: Character> '-' <token: Character>
+  ///     CustomCharClass -> Start Set (SetOp Set)* ']'
+  ///     Set             -> Member+
+  ///     Member          -> CustomCharClass | !']' !SetOp (Range | Atom)
+  ///     Range           -> Atom `-` Atom
   ///
-  mutating func parseCustomCharacterClass() throws -> CharacterClass {
-    try lexer.eat(expecting: .leftSquareBracket)
-    let isInverted = lexer.tryEat(.caret)
-    var components: [CharacterSetComponent] = []
-    while !lexer.tryEat(.rightSquareBracket) {
+  mutating func parseCustomCharacterClass(
+    _ start: CustomCharacterClass.Start
+  ) throws -> CustomCharacterClass {
+    typealias Member = CustomCharacterClass.Member
+    var members: Array<Member> = []
+
+    // TODO: Is this a correct/sane associativity? Precedence?
+    while true {
+      try source.expectNonEmpty()
+      try parseCCCMembers(into: &members)
```

Comment:
I believe this may be at a binary operator token if the previous iteration already lexed a binary operator

Reply:
Ah, I was thinking it would work out if `parseCCCMembers` was like `(!SetOp .)*`, that is it could consume zero or more non-set-operators. But I think I can simplify the loop logic a bit, since once we have a set op we're only ever iterating with the RHS.

Reply:
I went ahead and simplified this a bit, could you double check it's what you had in mind?